File: backend/app/models/coach.py
"""
AI Sales Coach "Luna" - Pydantic Models
TASK-029 / SPEC-028
"""

from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any, Literal
from datetime import datetime, time
from enum import Enum


# =============================================================================
# ENUMS
# =============================================================================

class EventType(str, Enum):
    PAGE_VIEW = "page_view"
    ACTION_COMPLETED = "action_completed"
    SUGGESTION_SHOWN = "suggestion_shown"
    SUGGESTION_CLICKED = "suggestion_clicked"
    SUGGESTION_DISMISSED = "suggestion_dismissed"
    SUGGESTION_SNOOZED = "suggestion_snoozed"
    WIDGET_EXPANDED = "widget_expanded"
    WIDGET_COLLAPSED = "widget_collapsed"
    SETTINGS_CHANGED = "settings_changed"


class SuggestionType(str, Enum):
    ADD_CONTACTS = "add_contacts"
    CREATE_PREP = "create_prep"
    CREATE_FOLLOWUP = "create_followup"
    GENERATE_ACTION = "generate_action"
    REVIEW_COACHING = "review_coaching"
    OVERDUE_PROSPECT = "overdue_prospect"
    MEETING_REMINDER = "meeting_reminder"
    COMPLETE_PROFILE = "complete_profile"
    TIP_OF_DAY = "tip_of_day"


class SuggestionAction(str, Enum):
    CLICKED = "clicked"
    DISMISSED = "dismissed"
    SNOOZED = "snoozed"
    EXPIRED = "expired"


class PatternType(str, Enum):
    WORK_HOURS = "work_hours"
    STEP_TIMING = "step_timing"
    PREFERRED_ACTIONS = "preferred_actions"
    DISMISS_PATTERNS = "dismiss_patterns"
    SUCCESS_PATTERNS = "success_patterns"


class NotificationFrequency(str, Enum):
    MINIMAL = "minimal"
    NORMAL = "normal"
    FREQUENT = "frequent"


class WidgetState(str, Enum):
    MINIMIZED = "minimized"
    COMPACT = "compact"
    EXPANDED = "expanded"
    HIDDEN = "hidden"


class EntityType(str, Enum):
    RESEARCH = "research"
    PREP = "prep"
    FOLLOWUP = "followup"
    PROSPECT = "prospect"
    DEAL = "deal"


# =============================================================================
# SETTINGS MODELS
# =============================================================================

class CoachSettingsBase(BaseModel):
    is_enabled: bool = True
    show_inline_tips: bool = True
    show_completion_modals: bool = True
    quiet_hours_start: Optional[str] = None  # HH:MM format
    quiet_hours_end: Optional[str] = None
    notification_frequency: NotificationFrequency = NotificationFrequency.NORMAL
    widget_state: WidgetState = WidgetState.MINIMIZED


class CoachSettingsUpdate(BaseModel):
    is_enabled: Optional[bool] = None
    show_inline_tips: Optional[bool] = None
    show_completion_modals: Optional[bool] = None
    quiet_hours_start: Optional[str] = None
    quiet_hours_end: Optional[str] = None
    notification_frequency: Optional[NotificationFrequency] = None
    widget_state: Optional[WidgetState] = None
    dismissed_tip_ids: Optional[List[str]] = None


class CoachSettings(CoachSettingsBase):
    id: str
    user_id: str
    dismissed_tip_ids: List[str] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# =============================================================================
# BEHAVIOR EVENT MODELS
# =============================================================================

class BehaviorEventCreate(BaseModel):
    event_type: EventType
    event_data: Dict[str, Any] = Field(default_factory=dict)
    page_context: Optional[str] = None


class BehaviorEvent(BaseModel):
    id: str
    user_id: str
    organization_id: str
    event_type: EventType
    event_data: Dict[str, Any]
    page_context: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True


# =============================================================================
# SUGGESTION MODELS
# =============================================================================

class SuggestionBase(BaseModel):
    suggestion_type: SuggestionType
    title: str
    description: str
    reason: Optional[str] = None
    priority: float = 50.0
    action_route: Optional[str] = None
    action_label: Optional[str] = None
    related_entity_type: Optional[EntityType] = None
    related_entity_id: Optional[str] = None
    icon: str = "💡"


class SuggestionCreate(SuggestionBase):
    expires_at: Optional[datetime] = None


class Suggestion(SuggestionBase):
    id: str
    user_id: str
    organization_id: str
    shown_at: datetime
    expires_at: Optional[datetime]
    action_taken: Optional[SuggestionAction]
    action_taken_at: Optional[datetime]
    snooze_until: Optional[datetime]
    feedback_rating: Optional[int]

    class Config:
        from_attributes = True


class SuggestionActionRequest(BaseModel):
    action: SuggestionAction
    snooze_until: Optional[datetime] = None
    feedback_rating: Optional[int] = Field(None, ge=1, le=5)


class SuggestionsResponse(BaseModel):
    suggestions: List[Suggestion]
    count: int
    has_priority: bool = False


# =============================================================================
# PATTERN MODELS
# =============================================================================

class UserPattern(BaseModel):
    id: str
    user_id: str
    organization_id: str
    pattern_type: PatternType
    pattern_data: Dict[str, Any]
    confidence: float
    sample_size: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class PatternsResponse(BaseModel):
    patterns: List[UserPattern]


# =============================================================================
# STATS MODELS
# =============================================================================

class TodayStats(BaseModel):
    research_completed: int = 0
    preps_completed: int = 0
    followups_completed: int = 0
    actions_generated: int = 0
    total_completed: int = 0
    streak_days: int = 0


class CoachStatsResponse(BaseModel):
    today: TodayStats
    suggestions_pending: int = 0
    patterns_learned: int = 0


# =============================================================================
# INSIGHT MODELS
# =============================================================================

class InsightType(str, Enum):
    CORRELATION = "correlation"
    PREDICTION = "prediction"
    RECOMMENDATION = "recommendation"
    TIP = "tip"


class Insight(BaseModel):
    id: str
    type: InsightType
    title: str
    description: str
    confidence: float = 0.5
    based_on: Optional[str] = None
    icon: str = "💡"
    action_route: Optional[str] = None
    action_label: Optional[str] = None


class InsightsResponse(BaseModel):
    insights: List[Insight]
    count: int


# =============================================================================
# CONTEXT MODEL (for rule evaluation)
# =============================================================================

class UserContext(BaseModel):
    """Context data used by the rule engine to generate suggestions."""
    user_id: str
    organization_id: str
    
    # Profile status
    has_sales_profile: bool = False
    has_company_profile: bool = False
    
    # Research data
    research_briefs: List[Dict[str, Any]] = Field(default_factory=list)
    research_without_contacts: List[Dict[str, Any]] = Field(default_factory=list)
    
    # Preparation data
    preps_without_followup: List[Dict[str, Any]] = Field(default_factory=list)
    preps_completed: List[Dict[str, Any]] = Field(default_factory=list)
    
    # Follow-up data
    followups_without_actions: List[Dict[str, Any]] = Field(default_factory=list)
    followups_completed: List[Dict[str, Any]] = Field(default_factory=list)
    
    # Prospects
    inactive_prospects: List[Dict[str, Any]] = Field(default_factory=list)
    
    # Patterns (learned)
    patterns: Dict[str, Any] = Field(default_factory=dict)
    
    # Time context
    current_hour: int = 12
    current_day_of_week: int = 0  # 0 = Monday


# =============================================================================
# RULE MODELS
# =============================================================================

class RuleDefinition(BaseModel):
    """Definition of a suggestion rule."""
    id: str
    name: str
    description: str
    suggestion_type: SuggestionType
    base_priority: float = 50.0
    icon: str = "💡"
    title_template: str  # Can include {company}, {name}, etc.
    description_template: str
    reason_template: Optional[str] = None
    action_route_template: Optional[str] = None
    action_label: Optional[str] = None

File: frontend/types/coach.ts
/**
 * AI Sales Coach "Luna" - TypeScript Types
 * TASK-029 / SPEC-028
 */

// =============================================================================
// ENUMS
// =============================================================================

export type EventType =
  | 'page_view'
  | 'action_completed'
  | 'suggestion_shown'
  | 'suggestion_clicked'
  | 'suggestion_dismissed'
  | 'suggestion_snoozed'
  | 'widget_expanded'
  | 'widget_collapsed'
  | 'settings_changed'

export type SuggestionType =
  | 'add_contacts'
  | 'create_prep'
  | 'create_followup'
  | 'generate_action'
  | 'review_coaching'
  | 'overdue_prospect'
  | 'meeting_reminder'
  | 'complete_profile'
  | 'tip_of_day'

export type SuggestionAction = 'clicked' | 'dismissed' | 'snoozed' | 'expired'

export type PatternType =
  | 'work_hours'
  | 'step_timing'
  | 'preferred_actions'
  | 'dismiss_patterns'
  | 'success_patterns'

export type NotificationFrequency = 'minimal' | 'normal' | 'frequent'

export type WidgetState = 'minimized' | 'compact' | 'expanded' | 'hidden'

export type EntityType = 'research' | 'prep' | 'followup' | 'prospect' | 'deal'


// =============================================================================
// SETTINGS
// =============================================================================

export interface CoachSettings {
  id: string
  user_id: string
  is_enabled: boolean
  show_inline_tips: boolean
  show_completion_modals: boolean
  quiet_hours_start: string | null
  quiet_hours_end: string | null
  notification_frequency: NotificationFrequency
  widget_state: WidgetState
  dismissed_tip_ids: string[]
  created_at: string
  updated_at: string
}

export interface CoachSettingsUpdate {
  is_enabled?: boolean
  show_inline_tips?: boolean
  show_completion_modals?: boolean
  quiet_hours_start?: string | null
  quiet_hours_end?: string | null
  notification_frequency?: NotificationFrequency
  widget_state?: WidgetState
  dismissed_tip_ids?: string[]
}


// =============================================================================
// SUGGESTIONS
// =============================================================================

export interface Suggestion {
  id: string
  user_id: string
  organization_id: string
  suggestion_type: SuggestionType
  title: string
  description: string
  reason: string | null
  priority: number
  action_route: string | null
  action_label: string | null
  icon: string
  related_entity_type: EntityType | null
  related_entity_id: string | null
  shown_at: string
  expires_at: string | null
  action_taken: SuggestionAction | null
  action_taken_at: string | null
  snooze_until: string | null
  feedback_rating: number | null
}

export interface SuggestionsResponse {
  suggestions: Suggestion[]
  count: number
  has_priority: boolean
}

export interface SuggestionActionRequest {
  action: SuggestionAction
  snooze_until?: string
  feedback_rating?: number
}


// =============================================================================
// BEHAVIOR EVENTS
// =============================================================================

export interface BehaviorEventCreate {
  event_type: EventType
  event_data?: Record<string, unknown>
  page_context?: string
}


// =============================================================================
// PATTERNS
// =============================================================================

export interface UserPattern {
  id: string
  user_id: string
  organization_id: string
  pattern_type: PatternType
  pattern_data: Record<string, unknown>
  confidence: number
  sample_size: number
  created_at: string
  updated_at: string
}

export interface PatternsResponse {
  patterns: UserPattern[]
}


// =============================================================================
// STATS
// =============================================================================

export interface TodayStats {
  research_completed: number
  preps_completed: number
  followups_completed: number
  actions_generated: number
  total_completed: number
  streak_days: number
}

export interface CoachStatsResponse {
  today: TodayStats
  suggestions_pending: number
  patterns_learned: number
}


// =============================================================================
// INLINE TIPS
// =============================================================================

export interface InlineTip {
  id: string
  title: string
  description: string
  icon: string
}

export interface InlineSuggestionsResponse {
  suggestions: InlineTip[]
  count: number
}


// =============================================================================
// CONTEXT
// =============================================================================

export interface CoachContextValue {
  // State
  isEnabled: boolean
  isLoading: boolean
  widgetState: WidgetState
  suggestions: Suggestion[]
  stats: CoachStatsResponse | null
  settings: CoachSettings | null
  
  // Actions
  setWidgetState: (state: WidgetState) => void
  refreshSuggestions: () => Promise<void>
  dismissSuggestion: (id: string) => Promise<void>
  snoozeSuggestion: (id: string, until: Date) => Promise<void>
  clickSuggestion: (id: string) => Promise<void>
  trackEvent: (type: EventType, data?: Record<string, unknown>) => Promise<void>
  updateSettings: (updates: CoachSettingsUpdate) => Promise<void>
  
  // Inline Tips
  dismissTipId: (tipId: string) => Promise<void>
  isDismissed: (tipId: string) => boolean
}


// =============================================================================
// SNOOZE OPTIONS
// =============================================================================

export interface SnoozeOption {
  label: string
  duration: number // minutes
}

export const SNOOZE_OPTIONS: SnoozeOption[] = [
  { label: '1 hour', duration: 60 },
  { label: 'Tonight', duration: getMinutesUntilTonight() },
  { label: 'Tomorrow', duration: getMinutesUntilTomorrow() },
]

function getMinutesUntilTonight(): number {
  const now = new Date()
  const tonight = new Date(now)
  tonight.setHours(20, 0, 0, 0) // 8 PM
  if (tonight <= now) {
    tonight.setDate(tonight.getDate() + 1)
  }
  return Math.ceil((tonight.getTime() - now.getTime()) / 60000)
}

function getMinutesUntilTomorrow(): number {
  const now = new Date()
  const tomorrow = new Date(now)
  tomorrow.setDate(tomorrow.getDate() + 1)
  tomorrow.setHours(9, 0, 0, 0) // 9 AM tomorrow
  return Math.ceil((tomorrow.getTime() - now.getTime()) / 60000)
}

